### ISSUE TITLE
Possibility to enable/disable an analysis by feature flag

### ISSUE THREAD

**Issue by xavijam** [CartoDB/cartodb#12025]
Right now we take the decision about offering an analysis (or not) if it is enabled by config in the case of it needs any service. Example, for data-observatory we have ([analyses.js file](https://github.com/CartoDB/cartodb/blob/master/lib/assets/core/javascripts/cartodb3/data/analyses.js#L80)):

```js
…
'data-observatory-measure': {
   …
    checkIfEnabled: function (configModel) {
      if (!configModel) throw new Error('configModel is required');
      return configModel.dataServiceEnabled('data_observatory');
    },
    …
  },
…
```

So, as you can see, we only check the availability with the configModel. One solution would be to add another parameter with the `userModel` for the `checkIfEnabled` function.

**Comment by javierarce** [comment]
Should we extend this new check to all the analysis or just the ones that are currently being checked by `dataServiceEnabled`?

**Comment by xavijam** [comment]
Well, if we pass it as an argument, and it is placed in the second position, we could "make it visible" unless it is need within the checker function.

```js
…
checkIfEnabled: function (configModel) {
   return configModel.get('whatever');
},

…
checkIfEnabled: function (configModel, userModel) {
   return userModel.hasFeatureFlagEnabled('whatever');
},

…
```

**Comment by javierarce** [comment]
For the feature flags I was thinking in using the same key we are using for the `dataServiceEnabled` check plus a prefix and a suffix:`analysis-` + key + `-enabled`.

So we could have:

- `analysis-data_observatory-enabled`
- `analysis-geocoder_internal-enabled`
- etc.

**Comment by xavijam** [comment]
Sorry but I don't get you. Why do you need to specify it right now? I mean, it should be added when it is developed right? not automatically...

**Comment by javierarce** [comment]
I thought that the original request was adding another check with the userModel, something like: 

```
…
checkIfEnabled: function (configModel, userModel) {
  return checkIfAnalysisIsEnabled('geocoder_internal', configModel, userModel);
},
…
```

```
var checkIfAnalysisIsEnabled = function (key, configModel, userModel) {
  if (!configModel) {
    throw new Error('configModel is required');
  }

  if (!userModel) {
    throw new Error('userModel is required');
  }

  return !!configModel.dataServiceEnabled(key) && userModel.featureEnabled(key);
};
```

**Comment by xavijam** [comment]
Yes and no. It depends on the analysis. For data-observatory we will need only to check if the "service" is enabled, that's all. In the future `nearest` analysis we will need to check only if the user has a feature flag in order to make it available...

**Comment by xavijam** [comment]
We will not create a feature flag for each analysis, that's not necessary.

**Comment by javierarce** [comment]
> Yes and no. It depends on the analysis. 

Oh, ok! So which analysis requires which method?

**Comment by xavijam** [comment]
For the moment none of the current analysis needs the "feature flag checker", but we will need it for the SQL function analysis that @ivanmalagon is developing right now (https://github.com/CartoDB/cartodb/issues/12018). That was the idea, just introduce the way to add it and check it before implementing the mentioned analysis.

**Comment by javierarce** [comment]
Mmm, I'm not sure if it makes sense to develop this as a separate feature before there's a real use for it.

**Comment by xavijam** [comment]
Well, there is one 💃 . We could develop this against @ivanmalagon branch, couldn't we?

**Comment by javierarce** [comment]
Oh, ok :)

**Comment by javierarce** [comment]
@ivanmalagon: in the `12018-sql-function` branch I see a `deprecated-sql-function` function, is that the one I could use to add the validation?

**Comment by ivanmalagon** [comment]
This issue has been merged in #12018 Next work to do on this will follow on that ticket.

**Comment by xavijam** [comment]
Great, thanks @ivanmalagon 💋 . Closing then!
